src/System.Private.CoreLib/shared/System/Runtime/InteropServices/Marshal.cs

```diff
@@ -0,0 +1,718 @@
+                {
+                    // unaligned read
+                    short val;
+                    byte* valPtr = (byte*)&val;
+                    valPtr[0] = addr[0];
+                    valPtr[1] = addr[1];
+                    return val;
+                }
+            }
+            catch (NullReferenceException)
+            {
+                // this method is documented to throw AccessViolationException on any AV
+                throw new AccessViolationException();
+            }
+        }
+
+        public static short ReadInt16(IntPtr ptr) => ReadInt16(ptr, 0);
+
+        public static unsafe int ReadInt32(IntPtr ptr, int ofs)
+        {
+            try
+            {
+                byte* addr = (byte*)ptr + ofs;
+                if ((unchecked((int)addr) & 0x3) == 0)
+                {
+                    // aligned read
+                    return *((int*)addr);
+                }
+                else
+                {
+                    // unaligned read
+                    int val;
+                    byte* valPtr = (byte*)&val;
+                    valPtr[0] = addr[0];
+                    valPtr[1] = addr[1];
+                    valPtr[2] = addr[2];
+                    valPtr[3] = addr[3];
+                    return val;
+                }
+            }
+            catch (NullReferenceException)
+            {
+                // this method is documented to throw AccessViolationException on any AV
+                throw new AccessViolationException();
+            }
+        }
+
+        public static int ReadInt32(IntPtr ptr) => ReadInt32(ptr, 0);
+
+        public static IntPtr ReadIntPtr(object ptr, int ofs)
+        {
+#if BIT64
+            return (IntPtr)ReadInt64(ptr, ofs);
+#else // 32
+            return (IntPtr)ReadInt32(ptr, ofs);
+#endif
+        }
+
+        public static IntPtr ReadIntPtr(IntPtr ptr, int ofs)
+        {
+#if BIT64
+            return (IntPtr)ReadInt64(ptr, ofs);
+#else // 32
+            return (IntPtr)ReadInt32(ptr, ofs);
+#endif
+        }
+
+        public static IntPtr ReadIntPtr(IntPtr ptr) => ReadIntPtr(ptr, 0);
+
+        public static unsafe long ReadInt64(IntPtr ptr, int ofs)
+        {
+            try
+            {
+                byte* addr = (byte*)ptr + ofs;
+                if ((unchecked((int)addr) & 0x7) == 0)
+                {
+                    // aligned read
+                    return *((long*)addr);
+                }
+                else
+                {
+                    // unaligned read
+                    long val;
+                    byte* valPtr = (byte*)&val;
+                    valPtr[0] = addr[0];
+                    valPtr[1] = addr[1];
+                    valPtr[2] = addr[2];
+                    valPtr[3] = addr[3];
+                    valPtr[4] = addr[4];
+                    valPtr[5] = addr[5];
+                    valPtr[6] = addr[6];
+                    valPtr[7] = addr[7];
+                    return val;
```

Comment:
For here and the other similar places, could we just use Unsafe.ReadUnaligned?

Reply:
`Unsafe.ReadUnaligned` would also make it work correctly on BigEndian platforms.

Reply:
Done